### PR TITLE
feat(routing): allow-offline with routing put

### PR DESCRIPTION
**Pull request by laurentsenta**
sister PR in kubo: https://github.com/ipfs/kubo/pull/9667

### CODE REVIEW

**Comment by Jorropo** [comment]
Is https://github.com/ipfs/kubo/pull/9667 correct ? I don't see how this relates

**Comment by laurentsenta** [comment]
@Jorropo yup: https://github.com/ipfs/kubo/pull/9667/commits/290f29afb33123d9d9bf98138b5a835942a56d3a

**Comment by BigLep** [comment]
@hacdias : can you please look at this again so we can get it merged?

**Review by hacdias — CHANGES_REQUESTED** [review]
Thanks @laurentsenta. This looks good to me.

I would like to see a test for this in [coreiface/tests/routing.go](https://github.com/ipfs/boxo/blob/main/coreiface/tests/routing.go) but I'm not sure how easy that would be. I think by using `tp.MakeAPISwarm(ctx, true, 1)` you'll be able to spin an offline node and therefore `Put` should work with `AllowOffline(true)` but not without. Note that this tests are run in Kubo's repo.

**Comment by laurentsenta** [comment]
@hacdias thanks for taking the time to review both PRs, 

From a quick look + tests:

- There is some coupling between online and full identity in the [node init](https://github.com/ipfs/kubo/blob/53821c202f14df2116459ba8729d61248342e15a/core/coreapi/test/api_test.go#L85)

I'll dig some more, but let me know if you've seen this before: when I try to start a node with full identity AND offline, I get a sigsev

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x40 pc=0x10131ab08]

goroutine 773 [running]:
github.com/ipfs/kubo/core/bootstrap.bootstrapRound({0x102732a98?, 0x14000a88640?}, {0x0, 0x0}, {0x14000315fb0?, 0x0?, 0x140003a6330?, 0x14000322fe0?})
	/Users/laurent/dev/plabs/ipfs/kubo/core/bootstrap/bootstrap.go:118 +0x68
github.com/ipfs/kubo/core/bootstrap.Bootstrap.func1({0x102741478?, 0x140004f5320?})
	/Users/laurent/dev/plabs/ipfs/kubo/core/bootstrap/bootstrap.go:89 +0xac
github.com/jbenet/goprocess.(*process).Go.func1()
	/Users/laurent/dev/go/pkg/mod/github.com/jbenet/goprocess@v0.1.4/impl-mutex.go:134 +0x3c
created by github.com/jbenet/goprocess.(*process).Go
	/Users/laurent/dev/go/pkg/mod/github.com/jbenet/goprocess@v0.1.4/impl-mutex.go:133 +0x204
FAIL	github.com/ipfs/kubo/core/coreapi/test	0.856s
FAIL
```

- Currently, the "sane" way to create an IPNS entry is by going through Publish + Get (that's what the test for Put does); that might be the right time to extract that code from the Publish function. We had a similar discussion about [verify and unmarshall](https://github.com/ipfs/gateway-conformance/pull/37#discussion_r1165450704).

**Comment by hacdias** [comment]
@laurentsenta I've never seen this error before, and it might be a bug? It seems like `host` is nil, which seems to make sense since we do not set `PeerHost` in `IpfsNode`. It seems like we'll need to make a few changes to the test function in Kubo to ensure we can have an offline and full identity node. But first, do we need full identity here?

> Currently, the "sane" way to create an IPNS entry is by going through Publish + Get (that's what the test for Put does); that might be the right time to extract that code from the Publish function.

Do you mean a `Create` function?

File: coreiface/tests/routing.go [hunk 1]
```diff
@@ -90,3 +97,36 @@ func (tp *TestSuite) TestRoutingPut(t *testing.T) {
...
+	err = api.Routing().Put(ctx, "/ipns/"+ipnsEntry.Name(), data)
+	if err == nil {
+		t.Fatal("this operation should fail because we are offline")
+	}
+
+	err = api.Routing().Put(ctx, "/ipns/"+ipnsEntry.Name(), data, options.Put.AllowOffline(true))
+	if err != nil {
+		t.Fatal(err)
+	}
+}
```
Comment (laurentsenta) [inline coreiface/tests/routing.go:127]: @hacdias I fixed the error by having a finer "isOnline" boolean in the `MakeAPI` code,

>> Currently, the "sane" way to create an IPNS entry is by going through Publish + Get (that's what the test for Put does); that might be the right time to extract that code from the Publish function.
>
> Do you mean a Create function?

Yes, see the code above I have to call `Publish`, then `Get`, just to get an IPNS payload that I can use with `Put`. If possible, I would have liked to be able to create the IPNS entry directly.

But seeing the result, it's less ugly than I thought, so maybe we're fine sticking with this approach. Let me know.

File: coreiface/tests/routing.go [hunk 2]
```diff
@@ -28,7 +30,12 @@ func (tp *TestSuite) testRoutingPublishKey(t *testing.T, ctx context.Context, ap
 		t.Fatal(err)
 	}
 
-	entry, err := api.Name().Publish(ctx, p)
+	// TODO: make sure we are fine with that implicit allow offline
```
Comment (hacdias) [inline coreiface/tests/routing.go]: Maybe you could add `options.NamePublishOption` in the arguments and let the caller decide. Just to be more flexible in the future.

File: coreiface/tests/routing.go (hunk 1 above)
Comment (hacdias) [reply coreiface/tests/routing.go:127]: I think it's fine keeping this. Please just look at my comment below.

File: coreiface/tests/routing.go (hunk 2 above)
Comment (laurentsenta) [reply coreiface/tests/routing.go]: That make sense, fixed!

**Review by hacdias — APPROVED** [review]

**Pull request merged**